src/java/com/puppetlabs/http/client/impl/Metrics.java

```diff
@@ -0,0 +1,257 @@
+    public static final String URL_METHOD_NAMESPACE = "with-url-and-method";
+    public static final String ID_NAMESPACE = "with-metric-id";
+    public static final String BYTES_READ_STRING = "bytes-read";
+    public enum MetricType { BYTES_READ; }
+
+    public static String metricTypeString(Metrics.MetricType metricType) {
+        // currently this is the only metric type we have; in the future when
+        // there are multiple types this will do something more useful
+        return BYTES_READ_STRING;
+    }
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(Metrics.class);
+
+    synchronized private static ClientTimer getOrAddTimer(MetricRegistry metricRegistry,
+                                                          String name, ClientTimer newTimer) {
+        Metric timer = metricRegistry.getMetrics().get(name);
+
+        if ( timer == null ) {
+            return metricRegistry.register(name, newTimer);
+        } else if ( ClientTimer.class.isInstance(timer) ) {
+            return (ClientTimer) timer;
+        } else {
+            throw new IllegalArgumentException(name +
+                    " is already used for a different type of metric");
+        }
+    }
+
+    private static ArrayList<Timer.Context> startBytesReadMetricIdTimers(MetricRegistry registry,
+                                                                         String[] metricId) {
+        ArrayList<Timer.Context> timerContexts = new ArrayList<>();
+        for (int i = 0; i < metricId.length; i++) {
+            ArrayList<String> currentId = new ArrayList<>();
+            for (int j = 0; j <= i; j++) {
+                currentId.add(metricId[j]);
+            }
+            ArrayList<String> currentIdWithNamespace = new ArrayList<>();
+            currentIdWithNamespace.add(ID_NAMESPACE);
+            currentIdWithNamespace.addAll(currentId);
+            currentIdWithNamespace.add(BYTES_READ_STRING);
+            String metric_name = MetricRegistry.name(METRIC_NAMESPACE,
+                    currentIdWithNamespace.toArray(new String[currentIdWithNamespace.size()]));
+
+            ClientTimer timer = new ClientTimer(metric_name, currentId, MetricType.BYTES_READ);
+            timerContexts.add(getOrAddTimer(registry, metric_name, timer).time());
+        }
+        return timerContexts;
+    }
+
+    private static ArrayList<Timer.Context> startBytesReadUrlTimers(MetricRegistry registry,
+                                                                    HttpRequest request) {
+        ArrayList<Timer.Context> timerContexts = new ArrayList<>();
+        try {
+            final RequestLine requestLine = request.getRequestLine();
+            final URI uri = new URI(requestLine.getUri());
+
+            // if the port is not specified, `getPort()` returns -1
+            final String port = uri.getPort() == -1 ? "" : ":" + uri.getPort();
+            final String strippedUrl = uri.getScheme() + "://" + uri.getHost()
```

Comment:
Rather than manually re-constructing the strippedUrl like this, I wonder if instead you could get the parts of the old URI, use those to construct a new one, and toString the result.  Something like:

```java
final String strippedUrl = new URI(uri.getScheme(), null, uri.getHost(), uri.getPort(), uri.getPath(), null, null).toString();
```

I think you want `getPath()` instead of `getRawPath()` for the path segment if you go with this approach to avoid double percent encoding the path.